Objects

jdoliner opened the pull request.

derekchiang commented on src/client/pfs/pfs.proto:
```diff
@@ -352,4 +360,19 @@ service BlockAPI {
+  rpc GetObject(Object) returns (google.protobuf.BytesValue) {}
+  rpc GetTag(Tag) returns (google.protobuf.BytesValue) {}
+  rpc Compact(google.protobuf.Empty) returns (google.protobuf.Empty) {}
+}
```
In a discussion offline we talked about changing the API to something like:

```
message Object {
  string name = 1;
}

message PutObjectRequest {
  bytes value = 1;
}

message TagObjectRequest {
  Object from = 1;
  Object to = 2;
}

service ObjectAPI {
  rpc Put(stream PutObjectRequest) returns (Object) {}
  rpc Get(Object) returns (google.protobuf.BytesValue) {}
  rpc Tag(TagObjectRequest) returns (google.protobuf.Empty) {}
  rpc Compact(google.protobuf.Empty) returns (google.protobuf.Empty)
}
```

That way we can tag objects retroactively.

derekchiang replied:
Actually, it occurred to me that we also need an `Exists` API that simply checks for the existence of a tag.  This is useful for a job to check if a datum needs to be reprocessed.

jdoliner commented:
Alright, this is finally ready for another look. I think I addressed all your concerns from before, with one exception which merging `GetTag` and `GetObject` into a single request. I decided this might not be a great idea because it makes both lookups slower and makes both require requesting an object which doesn't exist. S3 seems to start responding a lot slower if you send too many requests for objects that don't exist so it seemed like it wasn't worth it.

derekchiang commented on src/server/object_bench_test.go:
```diff
@@ -0,0 +1,96 @@
+package server
```
What are the results of the benchmarks?

jdoliner replied:
I was able to get about 37 MB/s of upload speed, 40 MB/s download when the cache was empty and 120 MB/s down when the cache was already loaded. All this was from multiple clients running on a single pod.

derekchiang commented on src/server/pfs/cmds/cmds.go:
```diff
@@ -739,6 +739,32 @@ files into your Pachyderm cluster.
 		}),
 	}
 
```
Should these be exposed to `pachctl`?

jdoliner replied:
I think it might be good to have them there even though they're lower level. They came in handy for debugging stuff and I could definitely see wanting to have a user debug something with their `pachctl`, that's the only use case I can think of though.

jdoliner commented on src/client/client.go:
```diff
@@ -50,7 +54,7 @@ const DefaultMaxConcurrentStreams uint = 100
 var (
 	// MaxMsgSize is used to define the GRPC frame size
-	MaxMsgSize = 10 * 1024 * 1024
+	MaxMsgSize = 20 * 1024 * 1024
```
In benchmarks I got the best results with a chunk size of `10 MB`, we cut this value in half to account for extra overhead in the message so this change puts us on the observed sweet spot for best performance.

jdoliner commented on src/client/pfs.go:
```diff
@@ -438,6 +442,114 @@ func (c APIClient) ListBlock() ([]*pfs.BlockInfo, error) {
+}
+
+// TagObject applies a tag to an existing object.
+func (c APIClient) TagObject(hash string, tags ...string) error {
```
Hmm, not sure I follow, are you saying that the signature would be:

```
TagObject(tags ...string)
```

It seems like that would be a weird signature since you could call it in ways that didn't really make sense, ie with a single `tag`.

Implementation wise I still kind of like distinguishing `tags` and `hashes` from each other since they come from different places and I can imagine a number of weird bugs that could happen if we accidentally overwrote a hash with a tag.

jdoliner commented on src/client/pfs.go:
```diff
@@ -438,6 +442,114 @@ func (c APIClient) ListBlock() ([]*pfs.BlockInfo, error) {
+	return buffer.Bytes(), nil
+}
+
+// Compact forces compaction of objects.
```
There is not, and actually compaction doesn't really work with the caching layer at all here because it assumes immutability and to do compaction you need to move things around. I think I might rip out the compaction code before I merge this actually. However, I was thinking an optimization which would actually work is to make it so that you can write multiple objects in a single streaming request and on the back end those could easily be compacted into a single block. It seems like that would be pretty good for the case where there are several small files in the same commit. However in the interest of getting this merged into master so you guys can use it for the new pps stuff I think it's probably better if I do that as a separate PR.
